internal/agent/device/hook/conditions.go

```diff
@@ -0,0 +1,192 @@
+package hook
+
+import (
+	"fmt"
+	"regexp"
+	"slices"
+	"strconv"
+	"strings"
+
+	ignv3types "github.com/coreos/ignition/v2/config/v3_4/types"
+	api "github.com/flightctl/flightctl/api/v1alpha1"
+)
+
+func checkCondition(cond *api.HookCondition, actionContext *actionContext) (bool, error) {
+	if cond == nil {
+		return true, nil
+	}
+
+	conditionType, err := (*cond).Type()
+	if err != nil {
+		return false, err
+	}
+
+	switch conditionType {
+	case api.HookConditionTypeExpression:
+		expression, err := (*cond).AsHookConditionExpression()
+		if err != nil {
+			return false, err
+		}
+		return checkExpressionCondition(expression, actionContext), nil
+	case api.HookConditionTypePathOp:
+		pathOp, err := (*cond).AsHookConditionPathOp()
+		if err != nil {
+			return false, err
+		}
+		return checkPathOpCondition(pathOp, actionContext), nil
+	default:
+		return false, fmt.Errorf("unknown hook condition type %q", conditionType)
+	}
+}
+
+const (
+	hookExpressionConditionFmt string = `(\w+)\s*(==)\s*(\w+)`
+)
+
+var hookExpressionConditionRegex = regexp.MustCompile(hookExpressionConditionFmt)
+
+func checkExpressionCondition(cond api.HookConditionExpression, actionCtx *actionContext) bool {
+	match := hookExpressionConditionRegex.FindStringSubmatch(cond)
+	if match == nil {
+		return false
+	}
+	lhs, op, rhs := match[1], match[2], match[3]
+	switch lhs {
+	case "rebooted":
+		b, err := strconv.ParseBool(rhs)
+		if err != nil {
+			return false
+		}
+		return checkOp(actionCtx.systemRebooted, op, b)
+	default:
+	}
+	return false
+}
+
+func checkOp[T comparable](lhs T, op string, rhs T) bool {
+	switch op {
+	case "==":
+		return checkEquals(lhs, rhs)
+	default:
+		return false
+	}
+}
+
+func checkEquals[T comparable](a, b T) bool {
+	return a == b
+}
+
+func checkPathOpCondition(cond api.HookConditionPathOp, actionCtx *actionContext) bool {
+	resetCommandLineVars(actionCtx)
+
+	isPathToDir := len(cond.Path) > 0 && cond.Path[len(cond.Path)-1] == '/'
+	if isPathToDir {
+		return checkPathOpConditionForDir(cond, actionCtx)
+	} else {
+		return checkPathOpConditionForFile(cond, actionCtx)
+	}
+}
+
+// checkFileOpConditionForDir checks whether a specified operation (create, update, remove) has been performed
+// on any file in the tree rooted at the specified path.
+// As a side-effect, it populates the command line variables of the action context with the corresponding list of files.
+func checkPathOpConditionForDir(cond api.HookConditionPathOp, actionCtx *actionContext) bool {
+	// ensure dir paths end with a trailing slash, so we don't accidentally match a file with a similar prefix
+	if len(cond.Path) < 1 || cond.Path[len(cond.Path)-1] != '/' {
+		cond.Path += "/"
```

Comment:
super nitty: totally get the reasoning here but we are mutating the input which we could just copy and handle.